command/connect/expose/expose.go

```diff
@@ -0,0 +1,238 @@
+		return 1
+	}
+	svc, svcNamespace, err := create.ParseIntentionTarget(c.service)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Invalid service name: %s", err))
+		return 1
+	}
+
+	if c.ingressGateway == "" {
+		c.UI.Error("An ingress gateway service must be given via the -ingress-gateway flag.")
+		return 1
+	}
+	gateway, gatewayNamespace, err := create.ParseIntentionTarget(c.ingressGateway)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Invalid ingress gateway name: %s", err))
+		return 1
+	}
+
+	if c.portRaw == "" {
+		c.UI.Error("A port must be provided via the -port flag.")
+		return 1
+	} else {
+		c.port, err = strconv.Atoi(c.portRaw)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error parsing port: %s", err))
+			return 1
+		}
+	}
+
+	// First get the config entry for the ingress gateway, if it exists. Don't error if it's a 404 as that
+	// just means we'll need to create a new config entry.
+	conf, _, err := client.ConfigEntries().Get(api.IngressGateway, gateway, nil)
+	if err != nil && !strings.Contains(err.Error(), agent.ConfigEntryNotFoundErr) {
+		c.UI.Error(fmt.Sprintf("Error fetching existing ingress gateway configuration: %s", err))
+		return 1
+	}
+	if conf == nil {
+		conf = &api.IngressGatewayConfigEntry{
+			Kind:      api.IngressGateway,
+			Name:      gateway,
+			Namespace: gatewayNamespace,
+		}
+	}
+
+	// Make sure the flags don't conflict with existing config.
+	ingressConf, ok := conf.(*api.IngressGatewayConfigEntry)
+	if !ok {
+		// This should never happen
+		c.UI.Error(fmt.Sprintf("Config entry is an invalid type: %T", conf))
+		return 1
+	}
+
+	listenerIdx := -1
+	for i, listener := range ingressConf.Listeners {
+		// Make sure the service isn't already exposed in this gateway
+		for _, service := range listener.Services {
+			if service.Name == svc {
+				c.UI.Error(fmt.Sprintf("Service %q already exposed through listener with port %d", svc, listener.Port))
+				goto CREATE_INTENTION
+			}
+		}
+
+		// If there's already a listener for the given port, make sure the protocol matches.
+		if listener.Port == c.port {
+			listenerIdx = i
+			if listener.Protocol != c.protocol {
+				c.UI.Error(fmt.Sprintf("Listener on port %d already configured with conflicting protocol %q", listener.Port, listener.Protocol))
+				return 1
+			}
+		}
+	}
+
+	// Add a service to the existing listener for the port if one exists, or make a new listener.
+	if listenerIdx >= 0 {
+		ingressConf.Listeners[listenerIdx].Services = append(ingressConf.Listeners[listenerIdx].Services, api.IngressService{
+			Name:      svc,
+			Namespace: svcNamespace,
+		})
+	} else {
+		ingressConf.Listeners = append(ingressConf.Listeners, api.IngressListener{
+			Port:     c.port,
+			Protocol: c.protocol,
+			Services: []api.IngressService{
+				{
+					Name:      svc,
+					Namespace: svcNamespace,
+				},
+			},
+		})
+	}
+
+	// Write the updated config entry using a check-and-set, so it fails if the entry
+	// has been changed since we looked it up.
+	{
```

Comment:
What is the reason for the brackets here?

Reply:
This was to enable using the `goto`, since the compiler wouldn't allow it to skip over the declaration of a variable (`succeeded`) otherwise.

Reply:
Gotcha, did not know that was a thing!